From c8220360c00d05e514fa708805ca8e59711ad2f8 Mon Sep 17 00:00:00 2001
From: Xiao Chen <xiao@apache.org>
Date: Tue, 11 Oct 2016 17:05:00 -0700
Subject: [PATCH 1929/2748] HADOOP-13698. Document caveat for KeyShell when
 underlying KeyProvider does not delete a key.

(cherry picked from commit b84c4891f9eca8d56593e48e9df88be42e24220d)
(cherry picked from commit bc5a9931d7a19b729e34ecd283d63245f4967667)

Conflicts:
	hadoop-common-project/hadoop-common/src/site/markdown/CommandsManual.md

Change-Id: If60648d1086f07843de75906c4289495865b12c1
---
 .../src/site/apt/CommandsManual.apt.vm             |   33 ++++++++++++++++++--
 1 file changed, 30 insertions(+), 3 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/site/apt/CommandsManual.apt.vm b/hadoop-common-project/hadoop-common/src/site/apt/CommandsManual.apt.vm
index 104c570..b66a30a 100644
--- a/hadoop-common-project/hadoop-common/src/site/apt/CommandsManual.apt.vm
+++ b/hadoop-common-project/hadoop-common/src/site/apt/CommandsManual.apt.vm
@@ -178,6 +178,33 @@ User Commands
    {{{../../hadoop-mapreduce-client/hadoop-mapreduce-client-core/MapredCommands.html#job}
    <<<mapred job>>>}} instead.
 
+* <<<key>>>
+
+
+   Manage keys via the KeyProvider. For details on KeyProviders, see the {{{../hadoop-hdfs/TransparentEncryption.html}Transparent Encryption Guide}}.
+
+   Providers frequently require that a password or other secret is supplied. If the provider requires a password and is unable to find one, it will use a default password and emit a warning message that the default password is being used. If the <<<-strict>>> flag is supplied, the warning message becomes an error message and the command returns immediately with an error status.
+
+   NOTE: Some KeyProviders (e.g. org.apache.hadoop.crypto.key.JavaKeyStoreProvider) do not support uppercase key names.
+
+   NOTE: Some KeyProviders do not directly execute a key deletion (e.g. performs a soft-delete instead, or delay the actual deletion, to prevent mistake). In these cases, one may encounter errors when creating/deleting a key with the same name after deleting it. Please check the underlying KeyProvider for details.
+
+   Usage: <<<hadoop key <subcommand> [options]>>>
+
+*-------------------+-------------------------------------------------------+
+||COMMAND_OPTION    ||                   Description
+*-------------------+-------------------------------------------------------+
+| create <keyname> [-cipher <cipher>] [-size <size>] [-description <description>] [-attr <attribute=value>] [-provider <provider>] [-strict] [-help] | Creates a new key for the name specified by the <keyname> argument within the provider specified by the <<<-provider>>> argument. The <<<-strict>>> flag will cause the command to fail if the provider uses a default password. You may specify a cipher with the <<<-cipher>>> argument. The default cipher is currently "AES/CTR/NoPadding". The default keysize is 128. You may specify the requested key length using the <<<-size>>> argument. Arbitrary attribute=value style attributes may be specified using the <<<-attr>>> argument. <<<-attr>>> may be specified multiple times, once per attribute.
+*-------------------+-------------------------------------------------------+
+| roll <keyname> [-provider <provider>] [-strict] [-help] | Creates a new version for the specified key within the provider indicated using the <<<-provider>>> argument. The <<<-strict>>> flag will cause the command to fail if the provider uses a default password.
+*-------------------+-------------------------------------------------------+
+| delete <keyname> [-provider <provider>] [-strict] [-f] [-help] | Deletes all versions of the key specified by the <keyname> argument from within the provider specified by <<<-provider>>>. The <<<-strict>>> flag will cause the command to fail if the provider uses a default password. The command asks for user confirmation unless <<<-f>>> is specified.
+*-------------------+-------------------------------------------------------+
+| list [-provider <provider>] [-strict] [-metadata] [-help] | Displays the keynames contained within a particular provider as configured in core-site.xml or specified with the <<<-provider>>> argument. The <<<-strict>>> flag will cause the command to fail if the provider uses a default password. <<<-metadata>>> displays the metadata.
+*-------------------+-------------------------------------------------------+
+|-help | Prints usage of this command
+*-------------------+-------------------------------------------------------+
+
 * <<<pipes>>>
 
    Deprecated. Use
@@ -238,7 +265,7 @@ Administration Commands
 * <<<daemonlog>>>
 
    Get/Set the log level for a Log identified by a qualified class name in the daemon dynamically.
-   By default, the command sends a HTTP request, but this can be overridden by using argument `-protocol https` to send a HTTPS request.
+   By default, the command sends a HTTP request, but this can be overridden by using argument <<<-protocol https>>> to send a HTTPS request.
 
    Usage: <<<hadoop daemonlog -getlevel <host:port> <classname> [-protocol (http|https)] >>>
 
@@ -250,12 +277,12 @@ Administration Commands
 | -getlevel <host:port> <name> [-protocol (http\|https)] | Prints the log level of the daemon running at
                                | <host:port>. This command internally connects
                                | to http://<host:port>/logLevel?log=<name>.
-                               | The `-protocol` flag specifies the protocol for connection.
+                               | The <<<-protocol>>> flag specifies the protocol for connection.
 *------------------------------+-----------------------------------------------------------+
 |   -setlevel <host:port> <name> <level> [-protocol (http\|https)] | Sets the log level of the daemon
                                | running at <host:port>. This command internally
                                | connects to http://<host:port>/logLevel?log=<name>.
-                               | The `-protocol` flag specifies the protocol for connection.
+                               | The <<<-protocol>>> flag specifies the protocol for connection.
 *------------------------------+-----------------------------------------------------------+
 
    Example:
-- 
1.7.9.5

